// clubhouse-lib/build/index.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _TokenRequestFactory = require('./TokenRequestFactory');

var _TokenRequestFactory2 = _interopRequireDefault(_TokenRequestFactory);

var _FetchRequestPerformer = require('./FetchRequestPerformer');

var _FetchRequestPerformer2 = _interopRequireDefault(_FetchRequestPerformer);

var _FetchRequestParser = require('./FetchRequestParser');

var _FetchRequestParser2 = _interopRequireDefault(_FetchRequestParser);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var API_BASE_URL = 'https://api.clubhouse.io';
var API_VERSION = 'beta';

/** */


var defaultConfig = {
  baseURL: API_BASE_URL,
  version: API_VERSION
};

/**
 * @class Client
 */

var Client = function () {
  function Client(requestFactory, requestPerformer, responseParser) {
    _classCallCheck(this, Client);

    this.requestFactory = requestFactory;
    this.requestPerformer = requestPerformer;
    this.responseParser = responseParser;
  }
  /** */


  _createClass(Client, [{
    key: 'listResource',
    value: function listResource(uri) {
      var request = this.requestFactory.createRequest(uri);
      return this.requestPerformer.performRequest(request).then(this.responseParser.parseResponse);
    }
  }, {
    key: 'getResource',
    value: function getResource(uri, params) {
      var request = params ? this.requestFactory.createRequest(uri, 'GET', params) : this.requestFactory.createRequest(uri);
      return this.requestPerformer.performRequest(request).then(this.responseParser.parseResponse);
    }
  }, {
    key: 'createResource',
    value: function createResource(uri, params) {
      var request = this.requestFactory.createRequest(uri, 'POST', params);
      return this.requestPerformer.performRequest(request).then(this.responseParser.parseResponse);
    }
  }, {
    key: 'updateResource',
    value: function updateResource(uri, params) {
      var request = this.requestFactory.createRequest(uri, 'PUT', params);
      return this.requestPerformer.performRequest(request).then(this.responseParser.parseResponse);
    }
  }, {
    key: 'deleteResource',
    value: function deleteResource(uri, params) {
      var request = this.requestFactory.createRequest(uri, 'DELETE', params);
      return this.requestPerformer.performRequest(request).then(this.responseParser.parseResponse);
    }

    /** */

  }, {
    key: 'listMembers',
    value: function listMembers() {
      return this.listResource('members');
    }

    /** */

  }, {
    key: 'getMember',
    value: function getMember(userID) {
      return this.getResource('members/' + userID);
    }

    /** */

  }, {
    key: 'getCurrentUser',
    value: function getCurrentUser() {
      return this.getResource('user');
    }

    /** */

  }, {
    key: 'listProjects',
    value: function listProjects() {
      return this.listResource('projects');
    }

    /** */

  }, {
    key: 'getProject',
    value: function getProject(id) {
      return this.getResource('projects/' + id);
    }

    /** */

  }, {
    key: 'createProject',
    value: function createProject(params) {
      return this.createResource('projects', params);
    }

    /** */

  }, {
    key: 'updateProject',
    value: function updateProject(id, params) {
      return this.updateResource('projects/' + id, params);
    }

    /** */

  }, {
    key: 'deleteProject',
    value: function deleteProject(id) {
      return this.deleteResource('projects/' + id);
    }

    /** */

  }, {
    key: 'listEpics',
    value: function listEpics() {
      return this.listResource('epics');
    }

    /** */

  }, {
    key: 'getEpic',
    value: function getEpic(epicID) {
      return this.getResource('epics/' + epicID);
    }

    /** */

  }, {
    key: 'createEpic',
    value: function createEpic(params) {
      return this.createResource('epics', params);
    }

    /** */

  }, {
    key: 'updateEpic',
    value: function updateEpic(epicID, params) {
      return this.updateResource('epics/' + epicID, params);
    }

    /** */

  }, {
    key: 'deleteEpic',
    value: function deleteEpic(epicID) {
      return this.deleteResource('epics/' + epicID);
    }

    /** */

  }, {
    key: 'addReaction',
    value: function addReaction(storyId, commentId, emoji) {
      return this.createResource('stories/' + storyId + '/comments/' + commentId + '/reactions', {
        emoji: emoji
      });
    }

    /** */

  }, {
    key: 'deleteReaction',
    value: function deleteReaction(storyId, commentId, emoji) {
      return this.deleteResource('stories/' + storyId + '/comments/' + commentId + '/reactions', {
        emoji: emoji
      });
    }

    /** */

  }, {
    key: 'listStories',
    value: function listStories(projectID) {
      return this.listResource('projects/' + projectID + '/stories');
    }

    /** */

  }, {
    key: 'searchStories',
    value: function searchStories(query, pageSize) {
      var _this = this;

      var processResult = function processResult(result) {
        if (result.next) {
          return _extends({}, result, {
            fetchNext: function fetchNext() {
              return _this.getResource(result.next).then(processResult);
            }
          });
        }
        return result;
      };
      return this.getResource('search/stories', {
        query: query,
        page_size: pageSize || 25
      }).then(processResult);
    }

    /** */

  }, {
    key: 'createStory',
    value: function createStory(params) {
      return this.createResource('stories', params);
    }

    /** */

  }, {
    key: 'getStory',
    value: function getStory(storyID) {
      return this.getResource('stories/' + storyID);
    }

    /** */

  }, {
    key: 'updateStory',
    value: function updateStory(storyID, params) {
      return this.updateResource('stories/' + storyID, params);
    }

    /** */

  }, {
    key: 'deleteStory',
    value: function deleteStory(storyID) {
      return this.deleteResource('stories/' + storyID);
    }

    /** */

  }, {
    key: 'createStoryComment',
    value: function createStoryComment(storyID, text) {
      return this.createResource('stories/' + storyID + '/comments', { text: text });
    }

    /** */

  }, {
    key: 'deleteStoryComment',
    value: function deleteStoryComment(storyID, commentID) {
      return this.deleteResource('stories/' + storyID + '/comments/' + commentID);
    }

    /** */

  }, {
    key: 'updateStoryComment',
    value: function updateStoryComment(storyID, commentID, text) {
      return this.updateResource('stories/' + storyID + '/comments/' + commentID, {
        text: text
      });
    }

    /** */

  }, {
    key: 'listTasks',
    value: function listTasks(storyID) {
      return this.listResource('stories/' + storyID + '/tasks');
    }

    /** */

  }, {
    key: 'createTask',
    value: function createTask(storyID, params) {
      return this.createResource('stories/' + storyID + '/tasks', params);
    }

    /** */

  }, {
    key: 'getTask',
    value: function getTask(storyID, taskID) {
      return this.getResource('stories/' + storyID + '/tasks/' + taskID);
    }

    /** */

  }, {
    key: 'updateTask',
    value: function updateTask(storyID, taskID, params) {
      return this.updateResource('stories/' + storyID + '/tasks/' + taskID, params);
    }

    /** */

  }, {
    key: 'deleteTask',
    value: function deleteTask(storyID, taskID) {
      return this.deleteResource('stories/' + storyID + '/tasks/' + taskID);
    }

    /** */

  }, {
    key: 'listWorkflows',
    value: function listWorkflows() {
      return this.listResource('workflows');
    }

    /** */

  }, {
    key: 'createStoryLink',
    value: function createStoryLink(params) {
      return this.createResource('story-links', params);
    }

    /** */

  }, {
    key: 'getStoryLink',
    value: function getStoryLink(storyLinkID) {
      return this.getResource('story-links/' + storyLinkID);
    }

    /** */

  }, {
    key: 'deleteStoryLink',
    value: function deleteStoryLink(storyLinkID) {
      return this.deleteResource('story-links/' + storyLinkID);
    }

    /** */

  }, {
    key: 'listFiles',
    value: function listFiles() {
      return this.listResource('files');
    }

    /** */

  }, {
    key: 'updateFile',
    value: function updateFile(fileID, params) {
      return this.updateResource('files/' + fileID, params);
    }

    /** */

  }, {
    key: 'deleteFile',
    value: function deleteFile(fileID) {
      return this.deleteResource('files/' + fileID);
    }

    /** */

  }, {
    key: 'listLinkedFiles',
    value: function listLinkedFiles() {
      return this.listResource('linked-files');
    }

    /** */

  }, {
    key: 'createLinkedFile',
    value: function createLinkedFile(params) {
      return this.createResource('linked-files', params);
    }

    /** */

  }, {
    key: 'updateLinkedFile',
    value: function updateLinkedFile(linkedFileID, params) {
      return this.updateResource('linked-files/' + linkedFileID, params);
    }

    /** */

  }, {
    key: 'deleteLinkedFile',
    value: function deleteLinkedFile(linkedFileID) {
      return this.deleteResource('linked-files/' + linkedFileID);
    }

    /** */

  }, {
    key: 'createLabel',
    value: function createLabel(name, color) {
      return this.createResource('labels', { name: name, color: color });
    }

    /** */

  }, {
    key: 'listLabels',
    value: function listLabels() {
      return this.listResource('labels');
    }

    /** */

  }, {
    key: 'listTeams',
    value: function listTeams() {
      return this.listResource('teams');
    }

    /** */

  }, {
    key: 'getTeam',
    value: function getTeam(teamID) {
      return this.getResource('teams/' + teamID);
    }

    /** */

  }, {
    key: 'listIterations',
    value: function listIterations() {
      return this.listResource('iterations');
    }

    /** */

  }, {
    key: 'createIteration',
    value: function createIteration(params) {
      return this.createResource('iterations', params);
    }

    /** */

  }, {
    key: 'getIteration',
    value: function getIteration(iterationID) {
      return this.getResource('iterations/' + iterationID);
    }

    /** */

  }, {
    key: 'updateIteration',
    value: function updateIteration(iterationID, params) {
      return this.updateResource('iterations/' + iterationID, params);
    }

    /** */

  }, {
    key: 'deleteIteration',
    value: function deleteIteration(iterationID) {
      return this.deleteResource('iterations/' + iterationID);
    }
  }], [{
    key: 'create',
    value: function create(token) {
      var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultConfig;
      var baseURL = config.baseURL,
          version = config.version;

      return new Client(new _TokenRequestFactory2.default(token, baseURL, version), new _FetchRequestPerformer2.default(), new _FetchRequestParser2.default());
    }
  }]);

  return Client;
}();

module.exports = Client;
