// clubhouse-lib/build/TokenRequestFactory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _universalUrl = require('universal-url');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

require('fetch-everywhere');

var TokenRequestFactory = function () {
  function TokenRequestFactory(token, baseURL, version) {
    _classCallCheck(this, TokenRequestFactory);

    this.token = token;
    this.baseURL = baseURL;
    this.version = version;
  }

  _createClass(TokenRequestFactory, [{
    key: 'prefixURI',
    value: function prefixURI(uri) {
      var prefix = '/api/' + this.version + '/';
      if (uri.startsWith(prefix)) {
        return uri;
      }
      return '' + prefix + uri;
    }
  }, {
    key: 'createRequest',
    value: function createRequest(uri) {
      var method = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'GET';
      var body = arguments[2];

      var url = new _universalUrl.URL(this.prefixURI(uri), this.baseURL);
      var headers = {
        Accept: 'application/json',
        'Content-Type': 'application/json; charset=utf-8'
      };

      url.searchParams.append('token', this.token);

      if (method === 'GET') {
        if (body) {
          Object.entries(body).forEach(function (entry) {
            return url.searchParams.append(entry[0], String(entry[1]));
          });
        }
        return new Request(url.toString(), {
          headers: headers,
          method: method
        });
      }

      return new Request(url.toString(), {
        body: JSON.stringify(body),
        headers: headers,
        method: method
      });
    }
  }]);

  return TokenRequestFactory;
}();

exports.default = TokenRequestFactory;
